opsgenie/resource_opsgenie_incident_template.go

```diff
@@ -0,0 +1,227 @@
+		Update: resourceOpsgenieIncidentTemplateUpdate,
+		Delete: resourceOpsgenieIncidentTemplateDelete,
+		Importer: &schema.ResourceImporter{
+			State: func(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				idParts := strings.Split(d.Id(), "/")
+				if len(idParts) != 36 || idParts[1] == "" {
+					return nil, fmt.Errorf("unexpected format of ID (%q), expected incident_template_id", d.Id())
+				}
+				d.SetId(idParts[1])
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"message": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"tags": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"details": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"priority": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"impacted_services": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"stakeholder_properties": {
+				Type:     schema.TypeList,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enable": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  true,
+						},
+						"message": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"description": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validation.StringLenBetween(1, 15000),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceOpsgenieIncidentTemplateCreate(d *schema.ResourceData, meta interface{}) error {
+	client, err := incident.NewClient(meta.(*OpsgenieClient).client.Config)
+	if err != nil {
+		return err
+	}
+	createRequest := &incident.CreateIncidentTemplateRequest{
+		Name:                  d.Get("name").(string),
+		Message:               d.Get("message").(string),
+		Description:           d.Get("description").(string),
+		Tags:                  expandOpsgenieIncidentTemplateTags(d.Get("tags").(*schema.Set)),
+		Details:               expandOpsgenieIncidentTemplateDetails(d.Get("details").(map[string]interface{})),
+		Priority:              incident.Priority(d.Get("priority").(string)),
+		ImpactedServices:      expandOpsgenieIncidentTemplateImpactedServices(d.Get("impacted_services").(*schema.Set)),
+		StakeholderProperties: expandOpsgenieIncidentTemplateStakeholderProperties(d.Get("stakeholder_properties").([]interface{})),
+	}
+	result, err := client.CreateIncidentTemplate(context.Background(), createRequest)
+	if err != nil {
+		return err
+	}
+	d.SetId(result.IncidentTemplateId)
+	return nil
```

Comment:
shouldn't we call read function here ?

Reply:
We are not calling the read function here because as per API docs the read here will not return single incident template created by the calling user. Instead it will return all the incident templates with all the details about them to the user which is unnecessary. API of incident template doesn't have implementation of GET actually. it's LIST 😸

Reply:
yes but in the read functino we should iterate all of them and find by the id until the api has support get request

Reply:
ok. Done.